3.0/m_relaymsg.cpp

```diff
@@ -0,0 +1,218 @@
+        if (!channel->HasUser(user))
+        {
+            user->WriteNumeric(Numerics::CannotSendTo(channel, "You must be in the channel to use this command."));
+            return CMD_FAILURE;
+        }
+
+        // Check that target nick is not already in use
+        if (ServerInstance->FindNick(nick))
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "RELAYMSG spoofed nick is already in use");
+            return CMD_FAILURE;
+        }
+
+        // Make sure the nick doesn't include any core IRC characters (e.g. *, !)
+        // This should still be more flexible than regular nick checking - in particular
+        // we want to allow "/" and "~" for relayers
+        for (std::string::const_iterator x = nick.begin(); x != nick.end(); x++)
+        {
+            if (invalid_chars_map.test(static_cast<unsigned char>(*x)))
+            {
+                user->WriteNumeric(ERR_BADRELAYNICK, nick, "Invalid characters in spoofed nick");
+                return CMD_FAILURE;
+            }
+        }
+
+        // If the sender was a lcoal user, check that the target includes a nick separator
+        if (IS_LOCAL(user))
+        {
+            bool accepted = false;
+            for (std::string::const_iterator x = cap.nick_separators.begin(); x != cap.nick_separators.end(); x++) {
+                if (nick.find(*x) != std::string::npos) {
+                    accepted = true;
+                }
+            }
+            if (!accepted)
+            {
+                user->WriteNumeric(ERR_BADRELAYNICK, nick, InspIRCd::Format("Spoofed nickname must include one of the following separators: %s", cap.nick_separators.c_str()));
+                return CMD_FAILURE;
+            }
+        }
+
+        // Send the message to everyone in the channel
+        std::string fakeSource = GetFakeHostmask(nick);
+        ClientProtocol::Messages::Privmsg privmsg(fakeSource, channel, text);
+        // Tag the message as @relaymsg=<nick> so the sender can recognize it
+        privmsg.AddTag("relaymsg", &captag, user->nick);
+        channel->Write(ServerInstance->GetRFCEvents().privmsg, privmsg);
+
+        if (IS_LOCAL(user))
+        {
+            // Pass the message on to other servers
+            CommandBase::Params params;
+            params.push_back(channame);
+            params.push_back(nick);
+            params.push_back(":" + text);
+
+            ServerInstance->PI->SendEncapsulatedData("*", "RELAYMSG", params, user);
+        }
+
+        return CMD_SUCCESS;
+    };
+};
+
+class ModuleRelayMsg : public Module
+{
+    RelayMsgCap cap;
+    RelayMsgCapTag captag;
+    CommandRelayMsg cmd;
+
+public:
+    ModuleRelayMsg() :
+        cap(this),
+        captag(this, cap),
+        cmd(this, cap, captag)
+    {
+    }
+
+    void ReadConfig(ConfigStatus& status) CXX11_OVERRIDE
+    {
+        ConfigTag* tag = ServerInstance->Config->ConfValue("relaymsg");
+        cap.nick_separators = tag->getString("separators", "/");
```

Comment:
This should be ensured to be non-empty.

```suggestion
        cap.nick_separators = tag->getString("separators", "/", 1);
```